### PR TITLE
Add option to use each bean OODB

### PR DESCRIPTION
**Pull request by Lynesth**
The whole point of this commit is to have each bean remember, once dispensed/found/loaded, which OODB they belong to so that we can get their relations or store/trash them without having to switch database each time. This is related to #629.

It may be a bad idea, it may be badly implemented, and it will surely fail some tests which I'll fix.

Anyway, this does the following:

- Moves the Tag and Duplication managers inside the OODB object instead of the facade (with getters and setters for each)
- Adds an option (because BC) to use each bean's OODB instead of the currently selected one when storing, trashing, duplicating, and such, using R::setUseBeanOODB( TRUE ). _Don't like the name of this option, but it'll do for now_
- In order for the previous bullet to work, it adds a new public function to OODBBean : getBeanHelper()
- The SimpleFacadeBeanHelper now has a __construct and its getToolbox() and getExtractedToolbox() functions have been modified to return the correct toolbox if the option is used.

I'm not sure about how I did this. Maybe we should move the manager (Assoc, Tag and Dup) directly in the Toolbox.

I at least hope we can start a discussion on this, because I think it would be some neat feature to have.

Lyn.

### CODE REVIEW

**Comment by Lynesth** [comment]
Ok so it seems that it didn't break BC, even though it adds more code than necessary...

It still needs some tests to check that it actually does its job of handling multiple databases.
The only one that would fail for sure at the moment is exportAll() because the loop mecanism isn't inside the Facade function like storeAll or trashAll.

I'll check how those tests are actually made and write some myself.

**Comment by Lynesth** [comment]
I made a little something to test the functionnality using RedUnit. You can check it out here : https://gist.github.com/Lynesth/cedc91c768a80d9b46ba262219db6af3

**Comment by gabordemooij** [comment]
This is an interesting feature, however this PR is quite intrusive. I have to look into this. Would it not be possible to just modify R::addDatabase() so that you can instead of an OODB have a OODBPool and then have the OODBPool determine the correct OODB to use? Just an idea, I might very well overlook some issues here..

**Comment by Lynesth** [comment]
I've been thinking about this whole thing and I don't understand why this shouldn't be the default behavior actually. Anyone would expect that, when storing or trashing a bean, it actually does it, whatever database you currently have selected... no ?

And I think that gathering all managers at the toolbox or the OODB level is the way to go.

Anyway, this PR is bad. It hasn't taken in account a lot of things, (passing Models instead of OODBBean to R::store() for example) so it's not good at all and would involve adding more logic in the Facace to make it work.

I'll try to think of some other way to do it, if you have any advice, let me know.

Lyn.

**Comment by gabordemooij** [comment]
What about this:

https://gist.github.com/gabordemooij/6feefc4964cafa766a085557d8900389

It seems to work, and requires almost no changes in RedBeanPHP at all.

File: RedBeanPHP/BeanHelper/SimpleFacadeBeanHelper.php [hunk 1]
```diff
@@ -31,6 +31,11 @@ class SimpleFacadeBeanHelper implements BeanHelper
 	 * @var \Closure
 	 */
 	private static $factory = null;
+	
+	/**
+	 * @var string
```
Comment (malukenho) [inline RedBeanPHP/BeanHelper/SimpleFacadeBeanHelper.php]: `string|null`

File: RedBeanPHP/BeanHelper/SimpleFacadeBeanHelper.php [hunk 2]
```diff
@@ -104,6 +126,11 @@ public function getModelForBean( OODBBean $bean )
 	 */
 	public function getExtractedToolbox()
 	{
-		return Facade::getExtractedToolbox();
+		if ( Facade::$useBeanOODB ) {
+			$toolbox = Facade::$toolboxes[$this->databaseKey];
+			return array( $toolbox->getRedbean(), $toolbox->getDatabaseAdapter(), $toolbox->getWriter(), $toolbox );
+		} else {
```
Comment (malukenho) [inline RedBeanPHP/BeanHelper/SimpleFacadeBeanHelper.php]: Do not need the `else` branch

File: RedBeanPHP/Facade.php [hunk 3]
```diff
@@ -68,7 +68,7 @@ class Facade
 	 * @var DBAdapter
 	 */
 	private static $adapter;
-
+
```
Comment (malukenho) [inline RedBeanPHP/Facade.php]: revert

File: RedBeanPHP/Facade.php [hunk 4]
```diff
@@ -525,7 +530,11 @@ public static function inspect( $type = NULL )
 	 */
 	public static function store( $bean )
 	{
-		return self::$redbean->store( $bean );
+		if (self::$useBeanOODB) {
+			return $bean->getBeanHelper()->getToolbox()->getRedBean()->store( $bean );
+		} else {
```
Comment (malukenho) [inline RedBeanPHP/Facade.php]: do not need the else branch

File: RedBeanPHP/Facade.php [hunk 5]
```diff
@@ -671,7 +680,11 @@ public static function loadForUpdate( $type, $id )
 	public static function trash( $beanOrType, $id = NULL )
 	{
 		if ( is_string( $beanOrType ) ) return self::trash( self::load( $beanOrType, $id ) );
-		return self::$redbean->trash( $beanOrType );
+		if ( self::$useBeanOODB ) {
+			return $beanOrType->getBeanHelper()->getToolbox()->getRedBean()->trash( $beanOrType );
+		} else {
```
Comment (malukenho) [inline RedBeanPHP/Facade.php]: same as above

**Comment by gabordemooij** [comment]
no feedback on this?

**Pull request closed without merging**